test/tstQueryTreeCallbackQueryPerThread.cpp

```diff
@@ -0,0 +1,113 @@
+/****************************************************************************
+ * Copyright (c) 2023 by the ArborX authors                                 *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#include "ArborX_EnableDeviceTypes.hpp" // ARBORX_DEVICE_TYPES
+#include <ArborX.hpp>
+
+#include "BoostTest_CUDA_clang_workarounds.hpp"
+#include <boost/test/unit_test.hpp>
+
+#include <numeric>
+#include <vector>
+
+BOOST_AUTO_TEST_SUITE(PerThread)
+
+struct IntersectionCallback
+{
+  int query_index;
+  bool &success;
+
+  template <typename Query, typename Value>
+  KOKKOS_FUNCTION void operator()(Query const &, Value const &value) const
+  {
+    success = (query_index == value.index);
+  }
+};
+
+BOOST_AUTO_TEST_CASE_TEMPLATE(callback_intersects, DeviceType,
+                              ARBORX_DEVICE_TYPES)
+{
+  using MemorySpace = typename DeviceType::memory_space;
+  using ExecutionSpace = typename DeviceType::execution_space;
+  using Tree = ArborX::BVH<MemorySpace>;
+
+  int const n = 10;
+  Kokkos::View<ArborX::Point *, DeviceType> points(
+      Kokkos::view_alloc(Kokkos::WithoutInitializing, "points"), n);
+  Kokkos::parallel_for(
+      Kokkos::RangePolicy<ExecutionSpace>(0, n), KOKKOS_LAMBDA(int i) {
+        points(i) = {{(float)i, (double)i, (double)i}};
+      });
+
+  Tree const tree(ExecutionSpace{}, points);
+
+  bool success;
+  Kokkos::parallel_reduce(
+      Kokkos::RangePolicy<ExecutionSpace>(0, n),
+      KOKKOS_LAMBDA(int i, bool &update) {
+        float center = i;
+        ArborX::Box box{{center - .5f, center - .5f, center - .5f},
+                        {center + .5f, center + .5f, center + .5f}};
+        tree.query(ArborX::Experimental::PerThread{}, ArborX::intersects(box),
+                   IntersectionCallback{i, update});
+      },
+      Kokkos::LAnd<bool, Kokkos::HostSpace>(success));
+
+  BOOST_TEST(success);
+}
+
+struct OrderedIntersectionCallback
+{
+  int query_index;
+  bool &success;
+
+  template <typename Query, typename Value>
+  KOKKOS_FUNCTION auto operator()(Query const &, Value const &value) const
+  {
+    success = (query_index == value.index);
+    return ArborX::CallbackTreeTraversalControl::early_exit;
```

Comment:
What is this testing?  Why are we doing early exit?

Reply:
The queries are executed in order of the distance. We are only interested in the first intersection that has zero distance.

Reply:
Looks to me that the predicate geometry is a box that will only contain a single point.

Reply:
There is only one primitive that will match what am I missing

Reply:
`OrderedIntersects` checks the distance for all primitives and executes them in order of increasing distance. For rays, this behavior corresponds to checking for intersections and that's why we chose the name. In this case, we would really process all primitives for all predicates (since all distances are finite).

Reply:
@dalg24 Did my explanation clarify why we need `early_exit` here?

Reply:
No it still does not make sense to me.  You are discussing some possible use of the callback.  I am looking at the one concrete use below and there is a single intersection so order does not matter.  Simplify your callback or change your test.